client/sites/settings/v2/router.tsx

```diff
@@ -0,0 +1,150 @@
+import {
+	Outlet,
+	Router,
+	createLazyRoute,
+	createMemoryHistory,
+	createRootRoute,
+	createRoute,
+	redirect,
+} from '@tanstack/react-router';
+import { siteSettingsQuery } from 'calypso/dashboard/app/queries';
+import { queryClient } from 'calypso/dashboard/app/query-client';
+
+const rootRoute = createRootRoute( { component: () => <Outlet /> } );
+
+const dashboardSiteSettingsCompatibilityRouteRoot = createRoute( {
+	getParentRoute: () => rootRoute,
+	path: 'sites/$siteSlug/settings',
+	loader: ( { params: { siteSlug } } ) => {
+		throw redirect( { to: `/${ siteSlug }` } );
+	},
+} );
+
+const dashboardSiteSettingsCompatibilityRouteWithFeature = createRoute( {
+	getParentRoute: () => rootRoute,
+	path: 'sites/$siteSlug/settings/$feature',
+	loader: ( { params: { siteSlug, feature } } ) => {
+		throw redirect( { to: `/${ siteSlug }/${ feature }` } );
+	},
+} );
+
+const siteRoute = createRoute( {
+	getParentRoute: () => rootRoute,
+	path: '$siteSlug',
+	loader: ( { params: { siteSlug } } ) =>
+		queryClient.ensureQueryData( siteSettingsQuery( siteSlug ) ),
+	component: () => <Outlet />,
+} );
+
+const settingsRoute = createRoute( {
+	getParentRoute: () => siteRoute,
+	path: '/',
+} ).lazy( () =>
+	import( 'calypso/dashboard/sites/settings' ).then( ( d ) =>
+		createLazyRoute( 'settings' )( {
+			component: () => <d.default siteSlug={ siteRoute.useParams().siteSlug } />,
+		} )
+	)
+);
+
+const siteVisibilityRoute = createRoute( {
+	getParentRoute: () => siteRoute,
+	path: 'site-visibility',
+	loader: ( { params: { siteSlug } } ) =>
+		queryClient.ensureQueryData( siteSettingsQuery( siteSlug ) ),
+} ).lazy( () =>
+	import( 'calypso/dashboard/sites/settings-site-visibility' ).then( ( d ) =>
+		createLazyRoute( 'site-visibility' )( {
+			component: () => <d.default siteSlug={ siteRoute.useParams().siteSlug } />,
+		} )
+	)
+);
+
+const subscriptionGiftingRoute = createRoute( {
+	getParentRoute: () => siteRoute,
+	path: 'subscription-gifting',
+	loader: ( { params: { siteSlug } } ) =>
+		queryClient.ensureQueryData( siteSettingsQuery( siteSlug ) ),
+} ).lazy( () =>
+	import( 'calypso/dashboard/sites/settings-subscription-gifting' ).then( ( d ) =>
+		createLazyRoute( 'subscription-gifting' )( {
+			component: () => <d.default siteSlug={ siteRoute.useParams().siteSlug } />,
+		} )
+	)
+);
+
+const createRouteTree = () =>
+	rootRoute.addChildren( [
+		siteRoute.addChildren( [ settingsRoute, siteVisibilityRoute, subscriptionGiftingRoute ] ),
+		dashboardSiteSettingsCompatibilityRouteRoot,
+		dashboardSiteSettingsCompatibilityRouteWithFeature,
+	] );
+
+const routeTree = createRouteTree();
```

Comment:
Lifting `routeTree` up out of the `getRouter()` function, just so that it can be used in a type definition, doesn't feel right to me. It doesn't seem right that the JS should have to change to support the TS.

```ts
const syncMemoryRouterToBrowserHistory = (
	router: Router< ReturnType< typeof createRouteTree > >
) => {
```

Reply:
Agreed. Let me see if using TanStack's AnyRouter type does the trick.